### PR TITLE
Add a limit for graph phrase query expansion

### PR DESCRIPTION
**Pull request by jimczi**
Backport of #34031 for 5.6.

### CODE REVIEW

**Review by jimczi — COMMENTED** [review]
@colings86 , as discussed with Jason yesterday I added a JVM option named `es.query.apply_graph_phrase_limit` to activate the limit in this version. By default it is deactivated and the only valid value for the JVM option is `true`.

File: core/src/main/java/org/apache/lucene/queryparser/classic/MapperQueryParser.java [hunk 1]
```diff
@@ -849,4 +856,118 @@ protected Query createFieldQuery(Analyzer analyzer, BooleanClause.Occur operator
...
+     * The JVM option can only be set to <code>true</code> (false is the default value), any other value
+     * will throw an {@link IllegalArgumentException}.
+     */
+    public static boolean shouldApplyGraphPhraseLimit() {
+        String value = System.getProperty("es.query.apply_graph_phrase_limit");
+        if (value == null) {
+            return false;
+        } else if (Booleans.parseBoolean(value, false) == false) {
+            throw new IllegalArgumentException("[" + value + "] is not a valid value for the JVM option:" +
+                "[es.query.apply_graph_phrase_limit]. Set it to [true] to activate the limit.");
```
Comment (colings86) [inline core/src/main/java/org/apache/lucene/queryparser/classic/MapperQueryParser.java]: Is there a way we can check this property on startup as well/instead? Maybe we could check it in SearchModule or MapperService?

The concern I see here is that restarts can be a pain for users and here they will only get feedback on the setting when they try to use a span query and then they can't use the query until they get the admin to restart the cluster

Comment (colings86) [reply core/src/main/java/org/apache/lucene/queryparser/classic/MapperQueryParser.java]: I also wonder if the user does not set the limit and we expand beyond the max clause count if we should emit a deprecation warning telling the user that this limit will be enforced in future versions?

**Review by colings86 — COMMENTED** [review]
@jimczi I left two comment on user feddback but otherwise I think this change is good

File: core/src/main/java/org/apache/lucene/queryparser/classic/MapperQueryParser.java (hunk 1 above)
Comment (jimczi) [reply core/src/main/java/org/apache/lucene/queryparser/classic/MapperQueryParser.java]: Good idea, I pushed https://github.com/elastic/elasticsearch/pull/34061/commits/5bd6406b49aa44d3b6f2d3e8fe6936232ff41042 to add the deprecation warning if the system property is not set. It also adds a validation of the system property (only `true` is valid) in `SearchModule`

File: core/src/main/java/org/apache/lucene/queryparser/classic/MapperQueryParser.java [hunk 2]
```diff
@@ -889,11 +888,12 @@ protected SpanQuery analyzeGraphPhrase(TokenStream source, String field, int phr
      * The JVM option can only be set to <code>true</code> (false is the default value), any other value
      * will throw an {@link IllegalArgumentException}.
      */
+    // public for tests
```
Comment (colings86) [inline core/src/main/java/org/apache/lucene/queryparser/classic/MapperQueryParser.java]: I think this isn't just public for tests now that `SearchModule` calls it?

**Review by colings86 — APPROVED** [review]
Left a super minor comment but this LGTM, thanks @jimczi

**Pull request merged**